src/Orleans.Core/Transactions/ITransactionInfo.cs

```diff
@@ -13,9 +16,18 @@ public interface ITransactionInfo
         string Id { get; }
 
         /// <summary>
-        /// Indicates that the transaction has aborted.
+        /// Record an unhandled exception thrown by a grain call. This causes the transaction to abort.
         /// </summary>
-        bool IsAborted { get; set; }
+        /// <param name="e">The unhandled exception that was thrown by the grain call</param>
+        /// <param name="serializationManager">The serialization manager used for serializing the exception</param>
+        void RecordException(Exception e, SerializationManager serializationManager);
```

Comment:
Why do we need to know the reason for the abort in the TransactionInfo?  Wasn't IsAborted sufficient?

Reply:
The point is that the user needs to know why a transaction aborted - so it can take appropriate action. For example, if the exception is caused by a timed out lock, it makes sense to retry the transaction... but it the user code throws an unexpected exception (e.g. IndexOutOfBounds) or expected exception (e.g. InsufficientFundsException), retrying is not what you want.

The reason why I felt this needed to be improved is that while implementing my general-purpose randomized serializability testing, I noticed that the exceptions returned are (1) much too big and bulky because of frequent re-throwing, and (2) terribly nested, basically unusable, if the transactions themselves are nested.

The solution I am proposing here guarantees that (no matter how much nesting or re-throwing happens) the user will always see either (1) a OrleansTransactionAbortedException with InnerException=(exception thrown by user code that caused the abort), if user code was the culprit, or (2) one of the specific exception classes describing the problem, if some issue within the transaction runtime was the culprit.

Reply:
Understood.  I'm not convinced this is a good way to address the error handling, but for the general abort logic I think this PR goes in the right direction.

Reuben, Would you object to letting the PR go in with a work item to revisit the exception handling, in a future pass?

I'm suggesting this becuase, while I don't agree with the current pattern, we should be able to revisit it with minimal effect to the application layer, the solution is non-trivial, and cleaning up this abort logic can help us narrow on the remaining issues.

In short, I'd rather get this improvement in then clean it up as opposed to holding it on how some (mostly) internal details are handled.

Reply:
Sounds fine to me